### PR TITLE
fix(preview): Close color palette when tabbed out of the palette focus

### PR DESCRIPTION
**Pull request by JChan106**
Fixes the bug where focusing out of the color palette using keyboard navigation didn't close the palette. The palette now closes when a user tabs out of the color palette, or if the user is tabbed in the palette and the user clicks out of the palette.

### CODE REVIEW

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 1]
```diff
@@ -19,15 +19,23 @@ export default function ColorPickerControl({
 }: Props): JSX.Element | null {
     const paletteRef = React.useRef<HTMLDivElement>(null);
     const [isColorPickerToggled, setIsColorPickerToggled] = React.useState(false);
-    const [isPaletteActive, handlers] = useAttention();
+    const { current: paletteEl } = paletteRef;
+
+    const handlePaletteBlur = ({ relatedTarget }: React.FocusEvent<HTMLButtonElement>): void => {
+        const targetElement = relatedTarget as HTMLButtonElement;
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:25]: I don't think IE11 supports `relatedTarget`, so we may need to fall back to `document.activeElement` here as in the `handleToggleBlur` function below.

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 2]
```diff
@@ -19,15 +19,23 @@ export default function ColorPickerControl({
...
+        if (targetElement.parentElement && targetElement.parentElement.parentElement !== paletteEl) {
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:26]: The `parentElement` calls here break encapsulation by making some assumptions about the DOM that may eventually change and break unexpectedly. Can we use `!paletteEl.contains(targetElement)`, instead?

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 3]
```diff
@@ -19,15 +19,23 @@ export default function ColorPickerControl({
...
+            setIsColorPickerToggled(false);
+        }
+    };
+
+    const [isPaletteActive, handlers] = useAttention({ onBlur: handlePaletteBlur });
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:31]: Can we add `onBlur` to `<ColorPickerPalette>` to avoid having to modify `useAttention` with separate callbacks?

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 4]
```diff
@@ -19,15 +19,23 @@ export default function ColorPickerControl({
...
 
     const handleSelect = (color: string): void => {
         setIsColorPickerToggled(false);
         onColorSelect(color);
     };
 
-    const handleBlur = ({ relatedTarget }: React.FocusEvent<HTMLButtonElement>): void => {
-        const { current: paletteEl } = paletteRef;
+    const handleToggleBlur = ({ relatedTarget }: React.FocusEvent<HTMLButtonElement>): void => {
```
[the remark this answers is not in the document]
Comment (JChan106) [reply src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:38]: Yeah just realized I can use the same handler if I take out the `isPaletteActive` conditional. I didn't see any focus issues in taking that conditional out.

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 5]
```diff
@@ -18,27 +18,31 @@ export default function ColorPickerControl({
     ...rest
 }: Props): JSX.Element | null {
     const paletteRef = React.useRef<HTMLDivElement>(null);
+    const toggleRef = React.useRef<HTMLButtonElement>(null);
     const [isColorPickerToggled, setIsColorPickerToggled] = React.useState(false);
-    const [isPaletteActive, handlers] = useAttention();
+    const [handlers] = useAttention();
+    const { current: paletteEl } = paletteRef;
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:24]: Can we revert this change? We typically destructure these inside the event handler where they're used, since the value could theoretically change before they're invoked.

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 6]
```diff
@@ -18,27 +18,31 @@ export default function ColorPickerControl({
...
         onColorSelect(color);
     };
 
     const handleBlur = ({ relatedTarget }: React.FocusEvent<HTMLButtonElement>): void => {
-        const { current: paletteEl } = paletteRef;
         // IE11 does not have relatedTarget but update activeElement before blur
         const nextTarget = relatedTarget || document.activeElement;
 
-        if (isPaletteActive || (nextTarget && paletteEl && paletteEl.contains(nextTarget as Node))) {
+        if ((nextTarget && paletteEl && paletteEl.contains(nextTarget as Node)) || toggleEl === (nextTarget as Node)) {
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:36]: Can we make this a bit more readable by using some additional variables?

```
const { current: paletteEl } = paletteRef;
const { current: toggleEl } = toggleRef;

// IE11 does not have relatedTarget but update activeElement before blur
const nextTarget = (relatedTarget || document.activeElement) as Node;
const isNextInPalette = paletteEl && paletteEl.contains(nextTarget);
const isNextToggle = toggleEl && toggleEl === nextTarget;

if (isNextInPalette || isNextToggle) {
    return;
}

setIsColorPickerToggled(false);

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 7]
```diff
@@ -18,27 +18,31 @@ export default function ColorPickerControl({
...
             return;
         }
 
         setIsColorPickerToggled(false);
     };
 
-    const handleClick = (): void => setIsColorPickerToggled(!isColorPickerToggled);
+    const handleClick = (): void => {
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:43]: Is this change needed or can we revert it?

File: src/lib/viewers/controls/color-picker/ColorPickerPalette.tsx [hunk 8]
```diff
@@ -3,10 +3,11 @@ import './ColorPickerPalette.scss';
 
 export type Props = {
     colors: Array<string>;
+    onBlur: React.FocusEventHandler<HTMLButtonElement>;
     onSelect: (color: string) => void;
 };
 
-export default function ColorPickerPalette({ colors, onSelect }: Props): JSX.Element {
+export default function ColorPickerPalette({ colors, onSelect, onBlur }: Props): JSX.Element {
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerPalette.tsx:10]: We typically alphabetize our props, attributes, etc. when possible.

File: src/lib/viewers/controls/color-picker/__tests__/ColorPickerControl-test.tsx [hunk 9]
```diff
@@ -76,6 +76,21 @@ describe('ColorPickerControl', () => {
             expect(getColorPickerPalette(wrapper).hasClass('bp-is-open')).toBe(true);
         });
 
+        test('should close the palette when focus is outside palette', () => {
+            const wrapper = getWrapper();
+            const toggleButton = getToggleButton(wrapper);
+            const divNode = document.createElement('div');
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/__tests__/ColorPickerControl-test.tsx:82]: We would typically refer to this as `divEl` or `divElement`, since we know it's an Element object rather than a generic Node.

File: src/lib/viewers/controls/color-picker/__tests__/ColorPickerControl-test.tsx [hunk 10]
```diff
@@ -76,6 +76,21 @@ describe('ColorPickerControl', () => {
...
+
+            toggleButton.simulate('click');
+            expect(getColorPickerPalette(wrapper).hasClass('bp-is-open')).toBe(true);
+
+            toggleButton.simulate('blur', {
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/__tests__/ColorPickerControl-test.tsx:87]: Should we be testing the palette instead of the button? Do we also need a test for when the `nextTarget` after blur is the toggle button?

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 11]
```diff
@@ -71,7 +76,12 @@ export default function ColorPickerControl({
                 data-testid="bp-ColorPickerControl-palette"
                 {...handlers}
             >
-                <ColorPickerPalette colors={colors} data-testid="bp-ColorPickerPalette" onSelect={handleSelect} />
+                <ColorPickerPalette
+                    colors={colors}
+                    data-testid="bp-ColorPickerPalette"
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:81]: This isn't related to your change, but does this get applied to the palette? It's not a listed prop and we don't have a `...rest` operator anywhere. Can we remove it safely?

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx [hunk 12]
```diff
@@ -18,27 +18,31 @@ export default function ColorPickerControl({
     ...rest
 }: Props): JSX.Element | null {
     const paletteRef = React.useRef<HTMLDivElement>(null);
+    const toggleRef = React.useRef<HTMLButtonElement>(null);
     const [isColorPickerToggled, setIsColorPickerToggled] = React.useState(false);
-    const [isPaletteActive, handlers] = useAttention();
+    const [handlers] = useAttention();
```
Comment (jstoffan) [inline src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:23]: Do we still need `useAttention` if we're not using the state it returns?

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx (hunk 7 above)
Comment (JChan106) [reply src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:43]: forgot that, thanks

File: src/lib/viewers/controls/color-picker/ColorPickerControl.tsx (hunk 12 above)
Comment (JChan106) [reply src/lib/viewers/controls/color-picker/ColorPickerControl.tsx:23]: Good point, will remove

**Review by jstoffan — APPROVED** [review]
